Combine ibus serial rx and telemetry on same uart and pin

unitware opened the pull request:
(This is a port of the now abandoned PR for cleanflight since CF went 2.x: https://github.com/cleanflight/cleanflight/pull/2554)

It's now possible to combine the serial rx and ibus telemetry on the same fc uart on a single pin.

I'm combining the ibus ports of the FS-iA6B receiver with a diode and a resistor.

```
+---------+
| FS-iA6B |
|         |
| Ser RX  |---|<---\       +------------+
|         |        |       | FC         |
| Sensor  |---[R]--*-------| Uart TX    |
+---------+                +------------+
```

R = 10Kohm, Diode 1N4148 or similar.

![image](https://cloud.githubusercontent.com/assets/6065378/23835685/0ed7bfc6-076c-11e7-9c27-dcf6b9686847.png)

Both uart tx and rx channels are used so it's not possible to use the spare pin for rx of something else.

Configure the serial port like this to get both serial rx and ibus telemetry on the same port:
```
serial 1 4160 115200 57600 115200 115200
```

It is still possible to run the serial rx and ibus telemetry on two uarts like before, an example:
```
serial 1 64 115200 57600 0 115200
serial 2 4096 115200 57600 115200 115200
```

Currently there is a problem (https://github.com/betaflight/betaflight/issues/2637) with ibus serial rx on master so I do not expect this PR to work at the moment but I have done a backport to 3.1.6 to test on hardware. There is also extensive unit tests to boost the confidence that this should be in good order once the aforementioned issue is solved.

In the meantime please give me feedback and why not test the backport: https://github.com/unitware/cleanflight/tree/bf-combined-ibus-on-3.1.6 

This is how the signals look:
![image](https://cloud.githubusercontent.com/assets/6065378/23835770/5dd5c9f0-076d-11e7-84eb-486759996cd2.png)

mikeller commented on src/main/target/NAZE/target.h:
```diff
@@ -111,6 +111,8 @@
 //#define SONAR_TRIGGER_PIN_PWM   PB8
 //#define SONAR_ECHO_PIN_PWM      PB9
 
+#define TELEMETRY_IBUS
```
This shouldn't be added to individual targets. Instead, add it to a class of targets here https://github.com/unitware/cleanflight/blob/154133f580024975a7fcf3b96a3dd6a95cb53b6f/src/main/target/common.h#L99

mikeller commented on src/main/telemetry/ibus_shared.c:
```diff
@@ -0,0 +1,309 @@
+ * clarify the protocol.
+ */
+
+/*
```
This should be converted to a wiki page, and added to https://github.com/betaflight/betaflight/wiki .

unitware replied:
https://github.com/betaflight/betaflight/pull/2639/commits/097b851c5ffe200fdcbe64916f10950f9b2fff49

unitware replied:
https://github.com/betaflight/betaflight/wiki/Ibus-telemetry

mikeller commented on src/main/telemetry/telemetry.h:
```diff
@@ -54,7 +54,7 @@ typedef struct telemetryConfig_s {
 PG_DECLARE(telemetryConfig_t, telemetryConfig);
 
-#define TELEMETRY_SHAREABLE_PORT_FUNCTIONS_MASK (FUNCTION_TELEMETRY_FRSKY | FUNCTION_TELEMETRY_LTM | FUNCTION_TELEMETRY_MAVLINK)
+#define TELEMETRY_SHAREABLE_PORT_FUNCTIONS_MASK (FUNCTION_TELEMETRY_FRSKY | FUNCTION_TELEMETRY_LTM | FUNCTION_TELEMETRY_MAVLINK | FUNCTION_TELEMETRY_IBUS)
```
Adding TELEMETRY_IBUS to this list makes it available for sharing with all supported serial RX protocols. Since sharing will only work when shared with  iBus RX, this should not be done. Instead, sharing between iBus RX and iBus telemetry should be enabled when iBus RX is enabled. See CRSF RX / CRSF telemetry for how it's done.

unitware replied:
ok, I see the problem and the proposed solution, is serial rx guaranteed to initialize before the telemetry?

mikeller replied:
At the moment it is: https://github.com/betaflight/betaflight/blob/master/src/main/fc/fc_init.c#L448-L480

I think there are other protocols like CRSF and JetiExBus that have the same requirements, so this is unlikely to change.

unitware replied:
I have made an (yet untested) to fix this in https://github.com/betaflight/betaflight/pull/2639/commits/5d828212419efc2249072241f4ddbb07134807f0 will see if it works tonight or tomorrow.

mikeller commented:
Is there a way to make this work on non-F3 MCUs (i.e. without the facility of having serial RX / TX on the same pin? Potentially with a similar solution to #1635.

unitware commented:
Thanks for the feedback so far. 

It should work on non bidirectional ports if rx and tx was connected with a diode but I'd like to verify the hardware before claiming it to be so.

unitware commented:
Yes, Ill do that and see if master is flyable later tonight

mikeller commented:
That's a good plan. There are some issues with USB on F4 at the moment, and if that makes it impossible for you to test I'd still be happy to merge this once it's rebased.

unitware commented:
ok, I have rebased and squashed, the strange issue https://github.com/betaflight/betaflight/issues/2637 remains and is not letting me test it more at this time.

unitware commented:
Thanks for your help and input @mikeller!

giacomo892 commented:
Hello! Thanks for your work @unitware . Do you mind briefly explaining why we need a resistor and a diode here?
Thanks

unitware commented:
It is a way to allow two outputs to share a single wire without shortcircuits. They are both high while idle.

The resistor must be on the telemetry pin since it is bidirectional. 

> 15 dec. 2017 kl. 19:07 skrev giacomo892 <notifications@github.com>:
> 
> Hello! Thanks for your work @unitware . Do you mind briefly explaining why we need a resistor and a diode here?
> Thanks
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

Peter010103 commented:
@unitware Hi, I bought [this receiver](https://www.banggood.com/Flit10-2_4G-10CH-Micro-Telemetry-Flysky-Compatible-Ibus-Receiver-for-FS-I6X-FS-i6S-Turnigy-Evolution-p-1260238.html?rmmds=search&cur_warehouse=CN) from BG, in hopes of using IBUS Telemetry. I've now hooked it up as you've shown, with the signal wire from the telemetry module to one of the TX ports. (TX3)

In the Betaflight's Port tab, I've selected the Telemetry Output as "IBUS" with baudrate 115200, but the FC does not reccognize the signal... What have I not done to get IBUS Telemtry working?

Thanks.

unitware commented:
no need for the CLI anymore, select telemetry output on the correct uart. on the ports tab. I guess you want serial rx on the same uart also. 

Follow the instructions on the BG link you linked above.

(I have not used that rx and telemetry module but it should work)
